pkg/webhook/pod/validating/pod_unavailable_budget.go

```diff
@@ -0,0 +1,170 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validating
+
+import (
+	"context"
+
+	"github.com/openkruise/kruise/pkg/control/pubcontrol"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/util/dryrun"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/apis/policy"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func (p *PodCreateHandler) podUnavailableBudgetValidatingPod(ctx context.Context, req admission.Request) (allowed bool, reason string, err error) {
+	allowed, newPod, oldPod, dryRun, err := p.shouldIgnore(ctx, req)
+	if err != nil || allowed {
+		return allowed, "", err
+	}
+
+	isUpdated := req.AdmissionRequest.Operation == admissionv1beta1.Update
+	// returns true for pod conditions that allow the operation for pod without checking PUB.
+	if newPod.Status.Phase == corev1.PodSucceeded || newPod.Status.Phase == corev1.PodFailed ||
+		newPod.Status.Phase == corev1.PodPending || !newPod.ObjectMeta.DeletionTimestamp.IsZero() {
+		klog.V(3).Infof("pod(%s.%s) Status(%s) Deletion(%v), then admit", newPod.Namespace, newPod.Name, newPod.Status.Phase, !newPod.ObjectMeta.DeletionTimestamp.IsZero())
+		return true, "", nil
+	}
+
+	pub, err := pubcontrol.GetPodUnavailableBudgetForPod(p.Client, p.controllerFinder, newPod)
+	if err != nil {
+		return false, "", err
+	}
+	// if there is no matching PodUnavailableBudget, just return true
+	if pub == nil {
+		return true, "", nil
+	}
+	control := pubcontrol.NewPubControl(pub)
+	klog.V(3).Infof("validating pod(%s.%s) operation(%s) for pub(%s.%s)", newPod.Namespace, newPod.Name, req.Operation, pub.Namespace, pub.Name)
+
+	// the change will not cause pod unavailability, then pass
+	if isUpdated && !control.IsPodUnavailableChanged(oldPod, newPod) {
+		klog.V(3).Infof("validate pod(%s.%s) changed cannot cause unavailability, then don't need check pub", newPod.Namespace, newPod.Name)
+		return true, "", nil
+	}
+
+	// when update operation, we should check whether old pod is ready
+	var checkPod *corev1.Pod
+	if isUpdated {
+		checkPod = oldPod
+	} else {
+		checkPod = newPod
+	}
+	return pubcontrol.PodUnavailableBudgetValidatePod(p.Client, checkPod, control, pubcontrol.Operation(req.Operation), dryRun)
+}
+
+func (p *PodCreateHandler) shouldIgnore(ctx context.Context, req admission.Request) (
+	// isIgnore indicates whether further logic need to be performed to check PUB
+	isIgnore bool, newPod, oldPod *corev1.Pod, dryRun bool, err error) {
+	isIgnore = true
+	// Ignore all calls to resources other than pods.
+	gvr := req.AdmissionRequest.Resource
```

Comment:
Is it possible that other resources are intercepted by the Pod validation webhook?